File: swift/Sources/Envoice/Invoice.swift
import Foundation

/// Result of an invoice generation
public enum InvoiceResult: Sendable {
    case success(InvoiceSuccess)
    case failure([ValidationError])

    public var isSuccess: Bool {
        if case .success = self { return true }
        return false
    }
}

/// Successful invoice generation result
public struct InvoiceSuccess: Sendable {
    public let pdfBase64: String
    public let filename: String
    public let validation: ValidationResult
    public let account: AccountInfo?

    /// Save the PDF to a file
    public func savePdf(to url: URL) throws {
        let data = try toData()
        try data.write(to: url)
    }

    /// Get the PDF as Data
    public func toData() throws -> Data {
        guard let data = Data(base64Encoded: pdfBase64) else {
            throw EnvoiceError.networkError("Failed to decode PDF")
        }
        return data
    }

    /// Get the PDF as a data URL
    public func toDataUrl() -> String {
        "data:application/pdf;base64,\(pdfBase64)"
    }
}

/// Fluent builder for creating invoices
public final class InvoiceBuilder: @unchecked Sendable {
    private let client: EnvoiceClient
    private var number: String?
    private var date: String?
    private var dueDate: String?
    private var seller: Party?
    private var buyer: Party?
    private var items: [LineItem] = []
    private var payment: PaymentInfo?
    private var currency = "EUR"
    private var template = "minimal"
    private var locale = "en"
    private var customization = Customization()

    init(client: EnvoiceClient) {
        self.client = client
    }

    /// Set the invoice number
    @discardableResult
    public func number(_ value: String) -> Self {
        self.number = value
        return self
    }

    /// Set the invoice date (YYYY-MM-DD)
    @discardableResult
    public func date(_ value: String) -> Self {
        self.date = value
        return self
    }

    /// Set the invoice date
    @discardableResult
    public func date(_ value: Date) -> Self {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withFullDate]
        self.date = formatter.string(from: value)
        return self
    }

    /// Set the due date
    @discardableResult
    public func dueDate(_ value: String) -> Self {
        self.dueDate = value
        return self
    }

    /// Set the seller
    @discardableResult
    public func seller(_ party: Party) -> Self {
        self.seller = party
        return self
    }

    /// Set the seller using closure
    @discardableResult
    public func seller(_ configure: (inout Party) -> Void) -> Self {
        var party = Party(name: "")
        configure(&party)
        self.seller = party
        return self
    }

    /// Set the buyer
    @discardableResult
    public func buyer(_ party: Party) -> Self {
        self.buyer = party
        return self
    }

    /// Set the buyer using closure
    @discardableResult
    public func buyer(_ configure: (inout Party) -> Void) -> Self {
        var party = Party(name: "")
        configure(&party)
        self.buyer = party
        return self
    }

    /// Add a line item
    @discardableResult
    public func addItem(_ item: LineItem) -> Self {
        self.items.append(item)
        return self
    }

    /// Add a line item using closure
    @discardableResult
    public func addItem(_ configure: (inout LineItem) -> Void) -> Self {
        var item = LineItem(description: "", quantity: 1, unitPrice: 0)
        configure(&item)
        self.items.append(item)
        return self
    }

    /// Set payment info
    @discardableResult
    public func payment(_ info: PaymentInfo) -> Self {
        self.payment = info
        return self
    }

    /// Set the currency
    @discardableResult
    public func currency(_ value: String) -> Self {
        self.currency = value
        return self
    }

    /// Set the template
    @discardableResult
    public func template(_ value: String) -> Self {
        self.template = value
        return self
    }

    /// Set the locale
    @discardableResult
    public func locale(_ value: String) -> Self {
        self.locale = value
        return self
    }

    /// Set logo from file URL
    @discardableResult
    public func logoFile(_ url: URL, widthMm: Int? = nil) throws -> Self {
        let data = try Data(contentsOf: url)
        customization.logoBase64 = data.base64EncodedString()
        customization.logoWidthMm = widthMm
        return self
    }

    /// Set logo from Base64
    @discardableResult
    public func logoBase64(_ base64: String, widthMm: Int? = nil) -> Self {
        customization.logoBase64 = base64
        customization.logoWidthMm = widthMm
        return self
    }

    /// Set footer text
    @discardableResult
    public func footerText(_ text: String) -> Self {
        customization.footerText = text
        return self
    }

    /// Set accent color
    @discardableResult
    public func accentColor(_ color: String) -> Self {
        customization.accentColor = color
        return self
    }

    /// Generate the invoice
    public func generate() async throws -> InvoiceResult {
        var errors: [ValidationError] = []

        if number == nil {
            errors.append(ValidationError(path: "$.invoice.number", code: "REQUIRED", message: "Invoice number is required"))
        }
        if date == nil {
            errors.append(ValidationError(path: "$.invoice.date", code: "REQUIRED", message: "Invoice date is required"))
        }
        if seller == nil {
            errors.append(ValidationError(path: "$.invoice.seller", code: "REQUIRED", message: "Seller information is required"))
        }
        if buyer == nil {
            errors.append(ValidationError(path: "$.invoice.buyer", code: "REQUIRED", message: "Buyer information is required"))
        }
        if items.isEmpty {
            errors.append(ValidationError(path: "$.invoice.items", code: "REQUIRED", message: "At least one line item is required"))
        }

        if !errors.isEmpty {
            return .failure(errors)
        }

        let hasCustomization = customization.logoBase64 != nil || customization.footerText != nil || customization.accentColor != nil

        let request = GenerateRequest(
            template: template,
            locale: locale,
            invoice: InvoiceData(
                number: number!,
                date: date!,
                dueDate: dueDate,
                seller: seller!,
                buyer: buyer!,
                items: items,
                payment: payment,
                currency: currency
            ),
            customization: hasCustomization ? customization : nil
        )

        return try await client.generateInvoice(request)
    }
}

File: kotlin/src/main/kotlin/dev/envoice/Invoice.kt
package dev.envoice

import java.io.File
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Base64

/**
 * Result of an invoice generation
 */
sealed class InvoiceResult {
    /**
     * Successful invoice generation
     */
    data class Success(
        val pdfBase64: String,
        val filename: String,
        val validation: ValidationResult,
        val account: AccountInfo? = null
    ) : InvoiceResult() {

        /**
         * Save the PDF to a file
         */
        fun savePdf(path: String) {
            savePdf(File(path))
        }

        /**
         * Save the PDF to a file
         */
        fun savePdf(file: File) {
            file.parentFile?.mkdirs()
            file.writeBytes(toBytes())
        }

        /**
         * Get the PDF as bytes
         */
        fun toBytes(): ByteArray = Base64.getDecoder().decode(pdfBase64)

        /**
         * Get the PDF as a data URL
         */
        fun toDataUrl(): String = "data:application/pdf;base64,$pdfBase64"
    }

    /**
     * Failed invoice generation with validation errors
     */
    data class Failure(val errors: List<ValidationError>) : InvoiceResult()
}

/**
 * Check if the result is successful
 */
val InvoiceResult.isSuccess: Boolean
    get() = this is InvoiceResult.Success

/**
 * Check if the result is a failure
 */
val InvoiceResult.isFailure: Boolean
    get() = this is InvoiceResult.Failure

/**
 * DSL builder for Party
 */
class PartyBuilder {
    var name: String = ""
    var street: String? = null
    var city: String? = null
    var postalCode: String? = null
    var country: String? = null
    var vatId: String? = null
    var email: String? = null
    var phone: String? = null

    fun build(): Party = Party(
        name = name,
        street = street,
        city = city,
        postalCode = postalCode,
        country = country,
        vatId = vatId,
        email = email,
        phone = phone
    )
}

/**
 * DSL builder for LineItem
 */
class LineItemBuilder {
    var description: String = ""
    var quantity: Double = 1.0
    var unit: String = "C62"
    var unitPrice: Double = 0.0
    var vatRate: Double = 19.0

    fun build(): LineItem = LineItem(
        description = description,
        quantity = quantity,
        unit = unit,
        unitPrice = unitPrice,
        vatRate = vatRate
    )
}

/**
 * Fluent builder for creating invoices
 */
class InvoiceBuilder internal constructor(private val client: EnvoiceClient) {
    private var number: String? = null
    private var date: String? = null
    private var dueDate: String? = null
    private var seller: Party? = null
    private var buyer: Party? = null
    private val items: MutableList<LineItem> = mutableListOf()
    private var payment: PaymentInfo? = null
    private var currency: String = "EUR"
    private var template: String = "minimal"
    private var locale: String = "en"
    private var logoBase64: String? = null
    private var logoWidthMm: Int? = null
    private var footerText: String? = null
    private var accentColor: String? = null

    /**
     * Set the invoice number
     */
    fun number(value: String): InvoiceBuilder {
        number = value
        return this
    }

    /**
     * Set the invoice date (ISO format: YYYY-MM-DD)
     */
    fun date(value: String): InvoiceBuilder {
        date = value
        return this
    }

    /**
     * Set the invoice date from LocalDate
     */
    fun date(value: LocalDate): InvoiceBuilder {
        date = value.format(DateTimeFormatter.ISO_LOCAL_DATE)
        return this
    }

    /**
     * Set the due date (ISO format: YYYY-MM-DD)
     */
    fun dueDate(value: String): InvoiceBuilder {
        dueDate = value
        return this
    }

    /**
     * Set the due date from LocalDate
     */
    fun dueDate(value: LocalDate): InvoiceBuilder {
        dueDate = value.format(DateTimeFormatter.ISO_LOCAL_DATE)
        return this
    }

    /**
     * Set the seller using DSL
     */
    fun seller(block: PartyBuilder.() -> Unit): InvoiceBuilder {
        seller = PartyBuilder().apply(block).build()
        return this
    }

    /**
     * Set the seller from Party object
     */
    fun seller(party: Party): InvoiceBuilder {
        seller = party
        return this
    }

    /**
     * Set the buyer using DSL
     */
    fun buyer(block: PartyBuilder.() -> Unit): InvoiceBuilder {
        buyer = PartyBuilder().apply(block).build()
        return this
    }

    /**
     * Set the buyer from Party object
     */
    fun buyer(party: Party): InvoiceBuilder {
        buyer = party
        return this
    }

    /**
     * Add a line item using DSL
     */
    fun addItem(block: LineItemBuilder.() -> Unit): InvoiceBuilder {
        items.add(LineItemBuilder().apply(block).build())
        return this
    }

    /**
     * Add a line item from LineItem object
     */
    fun addItem(item: LineItem): InvoiceBuilder {
        items.add(item)
        return this
    }

    /**
     * Set multiple items at once
     */
    fun items(items: List<LineItem>): InvoiceBuilder {
        this.items.clear()
        this.items.addAll(items)
        return this
    }

    /**
     * Set payment information
     */
    fun payment(info: PaymentInfo): InvoiceBuilder {
        payment = info
        return this
    }

    /**
     * Set payment information using builder
     */
    fun payment(
        iban: String? = null,
        bic: String? = null,
        terms: String? = null,
        reference: String? = null
    ): InvoiceBuilder {
        payment = PaymentInfo(iban, bic, terms, reference)
        return this
    }

    /**
     * Set the currency (default: EUR)
     */
    fun currency(value: String): InvoiceBuilder {
        currency = value
        return this
    }

    /**
     * Set the template style
     */
    fun template(value: String): InvoiceBuilder {
        template = value
        return this
    }

    /**
     * Set the locale
     */
    fun locale(value: String): InvoiceBuilder {
        locale = value
        return this
    }

    /**
     * Set logo from file
     */
    fun logoFile(path: String, widthMm: Int? = null): InvoiceBuilder {
        return logoFile(File(path), widthMm)
    }

    /**
     * Set logo from file
     */
    fun logoFile(file: File, widthMm: Int? = null): InvoiceBuilder {
        logoBase64 = Base64.getEncoder().encodeToString(file.readBytes())
        logoWidthMm = widthMm
        return this
    }

    /**
     * Set logo from Base64 string
     */
    fun logoBase64(base64: String, widthMm: Int? = null): InvoiceBuilder {
        logoBase64 = base64
        logoWidthMm = widthMm
        return this
    }

    /**
     * Set footer text
     */
    fun footerText(text: String): InvoiceBuilder {
        footerText = text
        return this
    }

    /**
     * Set accent color (hex code)
     */
    fun accentColor(color: String): InvoiceBuilder {
        accentColor = color
        return this
    }

    /**
     * Generate the invoice
     */
    suspend fun generate(): InvoiceResult {
        val errors = validateRequiredFields()
        if (errors.isNotEmpty()) {
            return InvoiceResult.Failure(errors)
        }

        val customization = if (logoBase64 != null || footerText != null || accentColor != null) {
            Customization(logoBase64, logoWidthMm, footerText, accentColor)
        } else null

        val request = GenerateRequest(
            template = template,
            locale = locale,
            invoice = InvoiceData(
                number = number!!,
                date = date!!,
                dueDate = dueDate,
                seller = seller!!,
                buyer = buyer!!,
                items = items.toList(),
                payment = payment,
                currency = currency
            ),
            customization = customization
        )

        return client.generateInvoice(request)
    }

    private fun validateRequiredFields(): List<ValidationError> {
        val errors = mutableListOf<ValidationError>()
        if (number == null) {
            errors.add(ValidationError("$.invoice.number", "REQUIRED", "Invoice number is required"))
        }
        if (date == null) {
            errors.add(ValidationError("$.invoice.date", "REQUIRED", "Invoice date is required"))
        }
        if (seller == null) {
            errors.add(ValidationError("$.invoice.seller", "REQUIRED", "Seller information is required"))
        }
        if (buyer == null) {
            errors.add(ValidationError("$.invoice.buyer", "REQUIRED", "Buyer information is required"))
        }
        if (items.isEmpty()) {
            errors.add(ValidationError("$.invoice.items", "REQUIRED", "At least one line item is required"))
        }
        return errors
    }
}
